keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,549 @@
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [API Proposal](#api-proposal)
+  - [Implementation](#implementation)
+  - [Validation](#validation)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+  - [Poor Person Workflow Engine](#poor-person-workflow-engine)
+  - [Increase number of patches to jobset](#increase-number-of-patches-to-jobset)
+- [Alternatives](#alternatives)
+  - [Operate on Conditions](#operate-on-conditions)
+<!-- /toc -->
+
+## Summary
+
+This KEP adds a StartupPolicy for Jobsets.  A StartupPolicy allows for some flexibility on when to start ReplicatedJobs in a Jobset.  
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+
+
+## Motivation
+
+High Performance Computing / Machine Learning users usually want some control over how jobs are being created.  A Startup policy would allow a user to specify which status a replicated job should have before starting other replicated jobs in a jobset.
+
+### Goals
+
+- Add a startup policy for JobSet
+
+### Non-Goals
+
+- Complicated workflow specifications
+  - This means that we will support simple dependecies rather than processing a full dag.
+  - Projects like argo-workflows are a better use case for this.
+- Starting Jobs based on Ready will be at the ReplicatedJob level not at the Job level.
+  - ReplicatedJobs will be classified as ready if all jobs are ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I have one ReplicatedJob that functions as the driver and a series of worker ReplicatedJobs.  
+I want the driver to be ready before I start the worker replicated jobs.  
+This is useful for HPC/AI/ML workloads as some frameworks have a driver pod that must be ready before the workers are started.  
```

Comment:
what happens to the second job if the first job gets recreated after all of them became ready?

Reply:
Good point.  I guess with this API I am not sure.  If the driver gets killed or preempted, I think the downstream objects should be stopped.

Reply:
If the launcher is essential to everything else, I would have it's (single pod) completion result in a failure in this case. So, no restarts and recreates unless the application stores state separately and is flexible to that.

Reply:
In my implementation PR, I had startup policy to be immutable so yes.

I'm not sure its worth the complication to allow for updates of the field.

Reply:
lets clearly state that then.

Reply:
I did have a section on validating saying that startupPolicy is immutable.

Reply:
It is not about mutability of the field, more about what we do across restarts, and here we are saying nothing.

We also need to clarify what we want to do across suspends, do we enforce the restart policy after resuming a suspended jobset?